### ISSUE TITLE
Restarting cri-o deletes and creates all running pods on fedora & ubuntu

### ISSUE THREAD

**Issue by ganeshmaharaj** [cri-o/cri-o#2479]
**Description**
Once cri-o has scheduled pods and they are running, if cri-o is restarted, all the pods sandboxes show as changed which leads to all of them getting killed and re-created. This issue was observed in Ubuntu and Fedora distros but not in ClearLinux or Core-os(minikube).

**Steps to reproduce the issue:**
1. Install kubernetes(1.14.x) and cri-o (1.14.x)
2. `kubeadm init`
3. Wait for pods to be up and running
4. `sudo systemctl restart crio`

**Describe the results you received:**
You will see under `crictl pods` that all the old pods are moved to `NotReady` state and kubelet schedules new pods.

**Describe the results you expected:**
cri-o should identify existing pods and not terminate them.
cri-o logs: https://gist.github.com/ganeshmaharaj/cd88dde27ae5b5827963bc36f30562ea
cri-o conf: https://gist.github.com/ganeshmaharaj/4ff9f50f7b92d3e63df830d6b83e1e12

**Additional information you deem important (e.g. issue happens only occasionally):**
K8s Version: 1.14.3
Cri-o version: 1.14.4, 1.14.1, master

**Output of `crio --version`:**
```
crio version 1.14.0-dev
commit: "4ccdc34243244cae7c4359991bb6bb02009daddd"

crio version 1.14.1
commit: ""
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
All testing was done on VMs for the different distros.

**Comment by ganeshmaharaj** [comment]
More logs --> http://paste.openstack.org/show/752898/

**Comment by ganeshmaharaj** [comment]
One more data point, systemd versions across the various distros.
ubuntu: 
```
systemd 237
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=hybrid
```

Fedora:
```
systemd 239
+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid
```

ClearLinux:
```
systemd 241 (241)
+PAM +AUDIT -SELINUX +IMA -APPARMOR -SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ -LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 -IDN -PCRE2 default-hierarchy=legacy
```

**Comment by haircommander** [comment]
can you post the clear linux crio.conf too?

**Comment by haircommander** [comment]
The one thing that stands out to me is the crio.conf was generated as if being called in userspace and not as root. Presumably that should work, but I am wondering if you're also calling crio from userspace in the other distros?

**Comment by ganeshmaharaj** [comment]
> can you post the clear linux crio.conf too?

https://github.com/cri-o/cri-o/issues/2479#issuecomment-501841766



> The one thing that stands out to me is the crio.conf was generated as if being called in userspace and not as root. Presumably that should work, but I am wondering if you're also calling crio from userspace in the other distros?

yes, it was created as the user and with the command `crio --config="" config --default > crio.conf`. It is later moved to `/etc/crio` and the service consumes it. Is that not an acceptable way of creating crio.conf?

**Comment by haircommander** [comment]
No it is definitely acceptable. I personally have had trouble running crio as a user, so I was just wondering.

**Comment by haircommander** [comment]
(or at least it's supposed to be acceptable :D )

**Comment by haircommander** [comment]
OOOH okay so the userspace config generated set storage_driver to vfs (which is really not that great). Can you regenerate the config as root, move it to /etc and try to reproduce?

**Comment by haircommander** [comment]
Also the config on clear linux seems to be generated using a newer crio

**Comment by ganeshmaharaj** [comment]
here is more logs. https://gist.github.com/ganeshmaharaj/fdfce5fe275ebd504195700a1042c5e9

**Comment by ganeshmaharaj** [comment]
clearlinux is using 1.14.1 while ubuntu is using head of master. I do not think clear would get a version that is later than master. 😕

**Comment by ganeshmaharaj** [comment]
`storage_driver = vfs` is commented out. I assume it will fall-back to the default and i was hoping that it would be overlay by default. Is that not true? just checking.

**Comment by haircommander** [comment]
if you're in userspace I think the default is vfs.

**Comment by ganeshmaharaj** [comment]
aah got it. In all the tests we are running the service via systemd. I just tested with creating `crio.conf` as root and explicitly setting `storage_driver = overlay` and same outcome.

**Comment by haircommander** [comment]
btw does this happen with runc and kata or just kata?

**Comment by ganeshmaharaj** [comment]
the runtime for all the pods on all distros is runc. The trigger is us trying to deploy kata and in the process restarting cri-o.

**Comment by haircommander** [comment]
you said clear linux is on 1.14.1? and ubuntu does the same with 1.14.1?

**Comment by haircommander** [comment]
can i have the clear linux logs?

**Comment by ganeshmaharaj** [comment]
> you said clear linux is on 1.14.1? and ubuntu does the same with 1.14.1?

yes.

> can i have the clear linux logs?

https://gist.github.com/ganeshmaharaj/c6bf583b67926552ea67ffc87ec430a8

**Comment by egernst** [comment]
/cc @mrunalp 
Anu updates on this issue?  Can folks on the cri-o side reproduce?

**Comment by ganeshmaharaj** [comment]
@haircommander anything else i can do here to help debug this issue?

**Comment by ganeshmaharaj** [comment]
I am able to reproduce the issue on fedora-30. 

Steps:
Fedora 30 VM, k8s (1.15.1), cri-o (1.14.9)
1. Install k8s with steps described at https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/
2. Install cri-o from source (make, make install, generate default config, copy systemd files to /etc/systemd/system and start service).
3. kubeadm init
4. Install CNI (flannel)
5. `kubectl apply -f https://raw.githubusercontent.com/kata-containers/packaging/master/kata-deploy/kata-rbac.yaml`
6. `kubectl apply -f https://raw.githubusercontent.com/kata-containers/packaging/master/kata-deploy/kata-deploy.yaml`
7. Watch crio restarting pods. output of crictl will show old pods are moved to `NotReady` state and new ones are spawned.

Ouput:
```
[vagrant@f-01 ~]$ sudo crictl -r unix:///run/crio/crio.sock pods
POD ID              CREATED             STATE               NAME                           NAMESPACE      ATTEMPT
5ed4b8198d3fa       3 seconds ago       Ready               kata-deploy-fhcc6              kube-system      1
a4783dad7cb40       3 seconds ago       Ready               coredns-5c98db65d4-jf76m       kube-system      1
b7e74d23f4af8       3 seconds ago       Ready               coredns-5c98db65d4-jwlq7       kube-system      1
2734e1f0ebddc       9 seconds ago       Ready               kube-scheduler-f-01            kube-system      1
d14978b383676       9 seconds ago       Ready               kube-controller-manager-f-01   kube-system      1
62deb1d7d43d4       9 seconds ago       Ready               etcd-f-01                      kube-system      1
fe43228f434d1       9 seconds ago       Ready               canal-qzmxt                    kube-system      1
367a14cc0917b       10 seconds ago      Ready               kube-proxy-8gp82               kube-system      1
6a61662941057       10 seconds ago      Ready               kube-apiserver-f-01            kube-system      1
ac42e9f0be854       36 seconds ago      NotReady            kata-deploy-fhcc6              kube-system      0
220e5b9447209       2 minutes ago       NotReady            coredns-5c98db65d4-jwlq7       kube-system      0
874f130765cde       2 minutes ago       NotReady            coredns-5c98db65d4-jf76m       kube-system      0
cc91fca6bded2       2 minutes ago       NotReady            kube-proxy-8gp82               kube-system      0
0c1a37d63e239       2 minutes ago       NotReady            canal-qzmxt                    kube-system      0
69738a4ffb1ec       2 minutes ago       NotReady            etcd-f-01                      kube-system      0
88c6a31bc5a9b       2 minutes ago       NotReady            kube-controller-manager-f-01   kube-system      0
2ec828087f79b       2 minutes ago       NotReady            kube-scheduler-f-01            kube-system      0
33feb83ab1add       2 minutes ago       NotReady            kube-apiserver-f-01            kube-system      0
```

**Comment by haircommander** [comment]
we determined this was due to running the node with cgroupfs cgroup-manager. This is because CRI-O is the parent process of the cgroup when using cgroupfs (as it is responsible for the creation and deletion). IIUC, when CRI-O restarts, the parent is gone, and the old containers have orphaned cgroups. the new CRI-O has to recreate all of the cgroups to properly be the parent of them.

We recommend systemd cgroup-manager, and set it as such by default. As such, I'm closing this.
